javalib/src/main/scala/java/math/BigDecimal.scala

```diff
@@ -672,13 +672,27 @@ class BigDecimal() extends Number with Comparable[BigDecimal] {
 
   def multiply(multiplicand: BigDecimal): BigDecimal = {
     val newScale = this._scale.toLong + multiplicand._scale
+
+    @inline
+    def bigMultiply: BigDecimal = {
+      val unscaled = this.getUnscaledValue.multiply(multiplicand.getUnscaledValue)
+      new BigDecimal(unscaled, safeLongToInt(newScale))
+    }
+
+    import java.lang.{Long => JLong}
     if (this.isZero || multiplicand.isZero) {
       zeroScaledBy(newScale)
-    } else if (this._bitLength + multiplicand._bitLength < 64) {
-      valueOf(this._smallValue * multiplicand._smallValue, safeLongToInt(newScale))
+    } else if (this._bitLength + multiplicand._bitLength >= 64) {
+      bigMultiply
     } else {
-      val unscaled = this.getUnscaledValue.multiply(multiplicand.getUnscaledValue)
-      new BigDecimal(unscaled, safeLongToInt(newScale))
+      val x = this._smallValue
+      val y = multiplicand._smallValue
+      val z = x * y
+      if (JLong.signum(z) == JLong.signum(x) * JLong.signum(y)) {
```

Comment:
I'll write up a proof -- I think the short answer is that checking the bit lengths guarantees you that the values aren't big enough to wrap back around. But I'll try to create a more formalized explanation for that.